src/Microsoft.VisualStudio.ProjectSystem.CSharp.VS/ProjectSystem/VS/MigrateXprojProjectFactory.cs

```diff
@@ -0,0 +1,142 @@
+// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using Microsoft.VisualStudio.Shell.Interop;
+using System;
+using Microsoft.VisualStudio.Packaging;
+using Microsoft.VisualStudio.Shell.Flavor;
+using System.Runtime.InteropServices;
+using System.Diagnostics;
+using System.IO;
+using System.Linq;
+
+namespace Microsoft.VisualStudio.ProjectSystem.VS.Xproj
+{
+    [Guid(CSharpProjectSystemPackage.XprojTypeGuid)]
+    internal sealed class MigrateXprojProjectFactory : FlavoredProjectFactoryBase, IVsProjectUpgradeViaFactory, IVsProjectUpgradeViaFactory4
+    {
+        public int UpgradeProject(string bstrFileName, uint fUpgradeFlag, string bstrCopyLocation, out string pbstrUpgradedFullyQualifiedFileName, IVsUpgradeLogger pLogger, out int pUpgradeRequired, out Guid pguidNewProjectFactory)
```

Comment:
We should probably add a test. It's probably going to need more mocks that it's worth it. Maybe split into smaller method chunks which are unit testable and leave all the VS specific piece out?